### PR TITLE
Release notes v0.42.0

### PR DESCRIPTION
**Pull request by codebien**

### CODE REVIEW

**Review by oleiade — DISMISSED** [review]
👏🏻 👏🏻 👏🏻

File: release notes/v0.42.0.md
Comment (mstoykov) [inline release notes/v0.42.0.md:9]:
```suggestion
Below is a simple example of using the new output in a `k6 run` command using the defaults options like the Remote Write server URL (http://localhost:9090/api/v1/write):
```

It seems strange to me to have the `O` uppercase and calling it "the output" makes it sound as it is the only output to me or like if it's something we didn't have anything now.

File: release notes/v0.42.0.md
Comment (mstoykov) [inline release notes/v0.42.0.md:7]:
```suggestion
Added a new builtin Output to any [Prometheus Remote Write](https://docs.google.com/document/d/1LPhVRSFkGNSuU1fBd81ulhsCPR4hkSZyyBj1SZ8fWOM/edit) implementations (e.g. [Prometheus](https://prometheus.io/docs/prometheus/latest/feature_flags/#remote-write-receiver), [Mimir](https://grafana.com/docs/mimir/latest/operators-guide/reference-http-api/#remote-write)). This is an experimental feature and it could change in the future introducing breaking changes.
```

The k6 parts is not an API arguably ;)

File: release notes/v0.42.0.md
Comment (mstoykov) [inline release notes/v0.42.0.md:15]:
```suggestion
It supports the new and convenient experimental [Native Histogram](https://prometheus.io/docs/practices/histograms) feature recently added in Prometheus >= `v2.40.0`. It is not enabled by default but we expect to make it the default way for mapping k6's [Trend metrics](https://k6.io/docs/javascript-api/k6-metrics/trend) when it is marked as mature enough from the Prometheus' project and supported by more Remote Write's implementation. For now, if you want to use it you need to set `K6_PROMETHEUS_RW_TREND_AS_NATIVE_HISTOGRAM` to `true`. 
```

File: release notes/v0.42.0.md
Comment (mstoykov) [inline release notes/v0.42.0.md:114]:
```suggestion
Thanks to the great effort from @eugercek, the [hosts](https://k6.io/docs/using-k6/k6-options/reference/#hosts) option now accepts domains that contain a wildcard at the beginning.
```

I think this is the only case where `*` is possible. Also maybe say whether that will map only 1 subdomain or mor, I don't remember that detail.

Maybe a sentence on what that can be used for will be nice.

File: release notes/v0.42.0.md
Comment (mstoykov) [inline release notes/v0.42.0.md:127]:
```suggestion
- [#2773](https://github.com/grafana/k6/pull/2773) Ensures that JavaScript runtime makes only one copy of the exports for each module including built-in ones.
```

File: release notes/v0.42.0.md
Comment (mstoykov) [inline release notes/v0.42.0.md:128]:
```suggestion
- [#2791](https://github.com/grafana/k6/pull/2791) Initialize VUS for `setup()` and `teardown()` only if they are defined in the script.
```

**Review by mstoykov — COMMENTED** [review]
great job :clap: :clap: . I have left some comments around the code.

File: release notes/v0.42.0.md
Comment (MattDodsonEnglish) [inline release notes/v0.42.0.md:7]:
```suggestion
This release brings a new builtin Output to any [Prometheus Remote Write](https://docs.google.com/document/d/1LPhVRSFkGNSuU1fBd81ulhsCPR4hkSZyyBj1SZ8fWOM/edit) implementations (e.g. [Prometheus](https://prometheus.io/docs/prometheus/latest/feature_flags/#remote-write-receiver), [Mimir](https://grafana.com/docs/mimir/latest/operators-guide/reference-http-api/#remote-write)). This is an experimental feature, and future releases could introduce breaking changes.
```

File: release notes/v0.42.0.md
Comment (MattDodsonEnglish) [inline release notes/v0.42.0.md:9]:
```suggestion
The following example uses `k6 run` with the new output. It uses the defaults options, such as the Remote Write server URL (http://localhost:9090/api/v1/write):
```

File: release notes/v0.42.0.md
Comment (MattDodsonEnglish) [inline release notes/v0.42.0.md:15]:
```suggestion
It supports the new and convenient experimental [Native Histogram](https://prometheus.io/docs/practices/histograms) feature, added in Prometheus `v2.40.0`. It's not enabled by default, but we expect to make it the default way to map k6 [Trend metrics](https://k6.io/docs/javascript-api/k6-metrics/trend) once the Prometheus' project signals that its mature enough and supported by more Remote Write's implementation. For now, if you want to use it, you need to set `K6_PROMETHEUS_RW_TREND_AS_NATIVE_HISTOGRAM` to `true`.
```

Comment (MattDodsonEnglish) [reply release notes/v0.42.0.md:15]: > supported by more Remote Write's implementation

I'm unclear about this one. Maybe

> ...when it supports more Remote Write implementations?

or 

> ...when more Remote Write implementations support it.

Something else?

File: release notes/v0.42.0.md
Comment (MattDodsonEnglish) [inline release notes/v0.42.0.md:17]:
```suggestion
You can find [complete documentation](https://k6.io/docs/results-output/real-time/prometheus-rw) with more examples, use cases, and available configurations.
```

File: release notes/v0.42.0.md
Comment (MattDodsonEnglish) [inline release notes/v0.42.0.md:26]:
```suggestion
<summary> Expand to see an example of the new WebSockets functionality</summary>
```

File: release notes/v0.42.0.md
Comment (MattDodsonEnglish) [inline release notes/v0.42.0.md:28]:
```suggestion
This example customizes tags for a WebScoket connection, sets up handlers using the new `on*` syntax, and demonstrates a `ping`/`pong` feature.
```

Comment (MattDodsonEnglish) [reply release notes/v0.42.0.md:28]: As a verb, it's "set up", and as a noun it's "setup". Annoying, I know :-(.

File: release notes/v0.42.0.md
Comment (MattDodsonEnglish) [inline release notes/v0.42.0.md:110]:
```suggestion
The [module docs](https://k6.io/docs/javascript-api/k6-experimental/websockets) have a complete reference, and some examples.
```

File: release notes/v0.42.0.md
Comment (MattDodsonEnglish) [inline release notes/v0.42.0.md:126]:
```suggestion
- [#2660](https://github.com/grafana/k6/pull/2660) Pre-loads the operating system TLS certificates. Thanks, @tbourrely.
```

File: release notes/v0.42.0.md
Comment (MattDodsonEnglish) [reply release notes/v0.42.0.md:128]: Perhaps "initializes" so that every item in the list has the same grammar.

File: release notes/v0.42.0.md
Comment (MattDodsonEnglish) [inline release notes/v0.42.0.md:132]:
```suggestion
- [#2712](https://github.com/grafana/k6/pull/2712) `k6/ws` returns an HTTP response for failed connections instead of an undefined behavior. Thanks, @brietaylor.
```

File: release notes/v0.42.0.md
Comment (MattDodsonEnglish) [inline release notes/v0.42.0.md:136]:
```suggestion
- [#2759](https://github.com/grafana/k6/pull/2759) Ensures the evaluation of thresholds over trend metrics' median.
```

File: release notes/v0.42.0.md
Comment (MattDodsonEnglish) [inline release notes/v0.42.0.md:137]:
```suggestion
- [#2759](https://github.com/grafana/k6/pull/2789) Fixes a few potential Output data races for interrupted test runs.
```

File: release notes/v0.42.0.md
Comment (MattDodsonEnglish) [inline release notes/v0.42.0.md:138]:
```suggestion
- [#2767](https://github.com/grafana/k6/pull/2767) Fixes the emission of `ws_session_duration` when setup throws an error.
```

File: release notes/v0.42.0.md
Comment (MattDodsonEnglish) [inline release notes/v0.42.0.md:149]:
```suggestion
- [#2791](https://github.com/grafana/k6/pull/2777) Optimized the performance of the internal Trie implementation. Thanks, @eugercek.
```

File: release notes/v0.42.0.md
Comment (MattDodsonEnglish) [inline release notes/v0.42.0.md:131]: Shouldn't we put this section at the top? I like the trend that @mstoykov started. I think it shows a lot of user respect that put breaks first, even if they're super minor.

**Review by MattDodsonEnglish — CHANGES_REQUESTED** [review]
Some nits, and I also prefer putting "Breaking changes" at the top, as we've started doing in the last few Release notes.

File: release notes/v0.42.0.md
Comment (codebien) [reply release notes/v0.42.0.md:128]:
```suggestion
- [#2791](https://github.com/grafana/k6/pull/2791) Initializes VUs for `setup()` and `teardown()` only if they are defined in the script.
```

File: release notes/v0.42.0.md
Comment (codebien) [reply release notes/v0.42.0.md]: For API I meant the k6 CLI's commands..for example removing the `experimental-` or adding the feature for setting the string URI. I will try to rephrase.

**Review by mstoykov — APPROVED** [review]

File: release notes/v0.42.0.md
Comment (imiric) [inline release notes/v0.42.0.md:27]:
```suggestion
It brings some useful features that the `k6/ws` module already has, like cookies, custom headers, compression and tags customization support, the syntax to define event handlers (`onopen`, `onmessage`, etc.) and `ping`/`pong` functionality.
```

File: release notes/v0.42.0.md
Comment (imiric) [inline release notes/v0.42.0.md:32]:
```suggestion
This example customizes tags for a WebSocket connection, sets up handlers using the new `on*` syntax, and demonstrates the `ping`/`pong` feature.
```

File: release notes/v0.42.0.md
Comment (imiric) [inline release notes/v0.42.0.md:44]: Hate to point out style issues :sweat_smile:, but this example doesn't use semi-colons and `var`/`let` consistently.

Let's use semi-colons everywhere, `const` where possible, and `let` everywhere else.

Also, maybe 2-space indentation, which is mostly the standard in JS. And, yes, we should update our sample scripts...

File: release notes/v0.42.0.md
Comment (imiric) [inline release notes/v0.42.0.md:119]: So that it matches the example.

```suggestion
It can be helpful for setting multiple subdomains of the same domain, so instead of setting `subdomain1.k6.io': '1.2.3.4', 'subdomain2.k6.io': '1.2.3.4'` it is possible to use the wildcard for setting directly `*.k6.io: '1.2.3.4'`.
```

File: release notes/v0.42.0.md
Comment (MattDodsonEnglish) [inline release notes/v0.42.0.md:2]:
```suggestion
k6 v0.42.0 is here!  :tada:

This release includes:
- A tiny breaking change to improve Websockets response handling.
- More features in our experimental modules
- Wildcard support for `hosts` 
- Some bug fixes, UX improvements, and maintenance
```

**Review by MattDodsonEnglish — APPROVED** [review]
I approve, but I'd prefer if we start with a quick TL;DR, so made a suggested intro.

I think the [v0.40 release](https://github.com/grafana/k6/blob/master/release%20notes/v0.40.0.md) set a good standard. If y'all accept this intro, I'm going to update the template.

File: release notes/v0.42.0.md
Comment (imiric) [inline release notes/v0.42.0.md:25]: Let's link to the xk6-websockets repo.

And we didn't do this in the v0.40.0 release notes, but we should explain _why_ we're introducing an experimental WS module that mostly mimics what `k6/ws` does.

We should explain this, mentioning what our goal for the experimental module is (will it eventually replace `k6/ws`?), what the differences are, and the current state of the experimental module. Users will probably be wondering when/if they should start using it over `k6/ws`.

Maybe some of this is a better fit for the documentation itself (I've yet to review https://github.com/grafana/k6-docs/pull/933/), but we should at least mention here what the current status of the project is after these updates.

**Review by imiric — CHANGES_REQUESTED** [review]

File: release notes/v0.42.0.md
Comment (codebien) [reply release notes/v0.42.0.md:2]:
```suggestion
k6 v0.42.0 is here!  :tada:

This release includes:
- A tiny breaking change to improve WebSockets response handling.
- A new experimental output.
- More features in our experimental WebSocket module.
- Wildcard support for `hosts`.
- Some bug fixes, UX improvements, and maintenance
```
@MattDodsonEnglish polished a bit more, WDYT?

Comment (MattDodsonEnglish) [reply release notes/v0.42.0.md:2]: LGTM!

File: release notes/v0.42.0.md
Comment (codebien) [reply release notes/v0.42.0.md]: >We should explain this, mentioning what our goal for the experimental module is (will it eventually replace k6/ws?), what the differences are, and the current state of the experimental module. Users will probably be wondering when/if they should start using it over k6/ws.

We should write this detailed info in the documentation then we could link it here. I think it is important enough to have it as permanent information and not in one specific release note. /cc @olegbespalov 

>Maybe some of this is a better fit for the documentation itself (I've yet to review https://github.com/grafana/k6-docs/pull/933), but we should at least mention here what the current status of the project is after these updates.

@olegbespalov can you help here?

**Review by andrewslotin — DISMISSED** [review]

File: release notes/v0.42.0.md
Comment (codebien) [reply release notes/v0.42.0.md]: It should be fixed by https://github.com/grafana/k6/pull/2811/commits/490cd786f1d3ebe201a2e425ea84bb080ff2b727

File: release notes/v0.42.0.md
Comment (olegbespalov) [inline release notes/v0.42.0.md:60]: Just to address @imiric 's requeest about the const
```suggestion
  const params = {
    "tags": {
      "my_tag": "hello"
    }
  };

  const ws = new WebSocket('ws://localhost:10000', null, params);
```

File: release notes/v0.42.0.md
Comment (imiric) [reply release notes/v0.42.0.md]: @olegbespalov I don't think we need to have a well-defined roadmap to make a short comment here about the project goals (replacing `k6/ws`?), and its current state (beyond just "experimental"; should people using `k6/ws` migrate to it, or people writing new WS scripts start using it instead of `k6/ws`?).

I think we missed the chance to mention the project goals when it was introduced in v0.40.0, so that can be done in the official documentation. But since we're announcing these updates here, we should at least expand on the status of it.

So maybe something like:

> This is still an experimental module, but with the recent changes we think it's usable for most users. So whether you're writing a new WebSocket test, or currently using the `k6/ws` module, we invite you to give it a try, and report any issues in the project's [issue tracker](https://github.com/grafana/xk6-websockets/issues).

Comment (olegbespalov) [reply release notes/v0.42.0.md]: >  I don't think we need to have a well-defined roadmap to make a short comment here about the project goals (replacing k6/ws?)

Not sure that it's the right time to commit that it replace the `k6/ws`, but the suggested sentence looks great :heart: , so I'm up to add this to the current release notes. 

Do you want me to add this, or wanna do this by yourself?

Comment (codebien) [reply release notes/v0.42.0.md]: @olegbespalov I've added it. I think this now conflicts a bit with the warning in the documentation which it is very alarming. For example:

>use at your own risk
 
If possible, we should relax and make it more welcoming.

Comment (olegbespalov) [reply release notes/v0.42.0.md]: > use at your own risk

It's a default warning that shows to all experimental extensions, I agree that it's a bit alarming, but probably it's better to tackle this separately, keeping in mind all the extensions.

Comment (olegbespalov) [reply release notes/v0.42.0.md]: But I've tried to use @imiric 's words magic to make the landing page moe welcoming, for now, hope that works :crossed_fingers:

File: release notes/v0.42.0.md
Comment (na--) [inline release notes/v0.42.0.md:142]:
```suggestion
### Enhancements and UX improvements
```

File: release notes/v0.42.0.md
Comment (na--) [inline release notes/v0.42.0.md:145]: Shouldn't this be in the bugfixes section?

**Review by na-- — DISMISSED** [review]
LGTM except the 2 minor nitpicks I left inline.

File: release notes/v0.42.0.md
Comment (codebien) [reply release notes/v0.42.0.md:145]: It seems it was an undefined behaviour before so I interpret it as an improvement. I defer to @mstoykov.

Comment (na--) [reply release notes/v0.42.0.md:145]: I am pretty sure it was a minor bug. Not so much "undefined behavior", rather it was "unintended" :smile:

**Review by imiric — DISMISSED** [review]
Great work everyone! :clap: :rocket: :shipit:

**Review by olegbespalov — APPROVED** [review]
:rocket:

**Review by oleiade — DISMISSED** [review]

File: release notes/v0.42.0.md
Comment (codebien) [inline release notes/v0.42.0.md:28]:
```suggestion
You can find [complete documentation](https://k6.io/docs/results-output/real-time/prometheus-remote-write) with more examples, use cases, and available configurations.
```

**Review by olegbespalov — APPROVED** [review]

**Review by na-- — APPROVED** [review]

**Pull request merged**
